common/src/backends/metrics.rs

```diff
@@ -33,3 +41,279 @@ where
         Ok(Metrics)
     }
 }
+
+type FnSpan = fn(&Request<Body>) -> Span;
+
+/// Record the tracing information for each request as given by the function to create a span
+pub struct TraceLayer<MakeSpan = MakeSpanSimple> {
+    fn_span: FnSpan,
+    make_span_type: PhantomData<MakeSpan>,
+}
+impl<MakeSpan> TraceLayer<MakeSpan> {
+    /// Create a trace layer using the give function to create spans. The span fields might be set by [Metrics] later.
+    ///
+    /// # Example
+    /// ```
+    /// TraceLayer::new(|request| {
+    ///     request_span!(
+    ///         request,
+    ///         request.params.param = field::Empty
+    ///     )
+    /// })
+    ///     .without_propagation()
+    ///     .build();
+    /// ```
+    pub fn new(fn_span: FnSpan) -> Self {
+        Self {
+            fn_span,
+            make_span_type: PhantomData,
+        }
+    }
+}
+
+impl<MakeSpan: tower_http::trace::MakeSpan<Body> + MakeSpanBuilder> TraceLayer<MakeSpan> {
+    /// Build the configured tracing layer
+    pub fn build(
+        self,
+    ) -> tower_http::trace::TraceLayer<
+        SharedClassifier<ServerErrorsAsFailures>,
+        MakeSpan,
+        DefaultOnRequest,
+        OnResponseStatusCode,
+    > {
+        tower_http::trace::TraceLayer::new_for_http()
+            .make_span_with(MakeSpan::new(self.fn_span))
+            .on_response(OnResponseStatusCode)
+    }
+}
+
+impl TraceLayer<MakeSpanSimple> {
+    /// Switch to the span maker which does not add propagation details from the request headers
+    pub fn without_propagation(self) -> Self {
+        self
+    }
+}
+
+impl TraceLayer<MakeSpanPropagation> {
+    /// Switch to the span maker which adds propagation details from the request headers
```

Comment:
Tracing noob question: it means propagation from request headers to response headers, right?

Reply:
Only from the request headers. If we have an HTTP client that calls another service (like gateway calling the deployers), then we have to manually set these propagation details on the headers (currently)

Reply:
Here's an example of the headers being set
https://github.com/shuttle-hq/shuttle/blob/b3e11b5ec0eb129dfbe06f90b99c785ea5334126/gateway/src/proxy.rs#L157-L159

Reply:
So `MakeSpanPropagation` will set it from the request headers to the tracing context (for the parent span) to be more precise

Reply:
Thanks!